new design note for gateway timeouts

superbeeny opened the pull request:
New design note for gateway timeouts

ytimocin commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+team. Do not provide the design details in this, section - there is a
+dedicated section for that later in the document.
+-->
+The purpose of this feature is to allow the user to configure a timeout on the gateway for an application. This will allow the user to specify how long the gateway should wait for a response from the application before timing out. This will be useful in scenarios where the application may take a long time to respond, or where the user wants to ensure that the application responds within a certain ...
```
What is the main difference between health/readiness probes and this? Is this for an endpoint that would take a long time to process something?

superbeeny replied:
Yes, my explicit use case is for LLM responses

willdavsmith commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+we plan to do in the future, but are out of scope of this design, list
+them here. Provide a brief explanation on why this is a non-goal.
+-->
+None
```
Is it out of scope to provide more httpproxy punch-through options? Is this pattern extensible to other routing configuration users may want to provide in the future?

superbeeny replied:
Will add note to doc to limit scope to requirements

willdavsmith commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+        destination: 'http://${ui.name}:3000'
+      }
+      {
+        {
```
syntax is kind of confusing. if I'm reading it right, this `timeoutPolicy` is applied to the `/api` route. why is `timeoutPolicy` not present at the top level along with `path` and `destination`?

superbeeny replied:
this is a typo that I will fix

nithyatsu commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+team. Do not provide the design details in this, section - there is a
+dedicated section for that later in the document.
+-->
+The purpose of this feature is to allow the user to configure a timeout on the gateway for an application. This will allow the user to specify how long the gateway should wait for a response from the application before timing out. This will be useful in scenarios where the application may take a long time to respond, or where the user wants to ensure that the application responds within a certain ...
```
Are the timeouts configurable per route?

superbeeny replied:
Yes

willdavsmith commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+      }
+      {
+        {
+          timeoutPolicy: {
```
Is there a way (or a need) to apply global timeout policies for all routes?

superbeeny replied:
Not necessary for this PR but worth some further investigation

kachawla commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+  **idle:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
+  **idleConnection:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
+
+This maps to the contour timeoutPolicy - https://github.com/projectcontour/contour/blob/main/apis/projectcontour/v1/httpproxy.go#L1126
```
Do we need to consider how this design will if we support other options in future?

willtsai replied:
+1 as removing Contour as a hard dependency is on the roadmap: https://github.com/radius-project/roadmap/issues/58

superbeeny replied:
This will need refactoring, will check with k8s gateway API for similarities

superbeeny replied:
Yes, I looked into this and adapted to follow the k8s gateway api spec

kachawla commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+
+The solution would update the specification for the gateway resource in the bicep file to include a timeoutPolicy object. This object would contain the following properties:
+  **response:** The amount of time the gateway should wait for a response from the application before timing out.
+  **idle:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
+  **idleConnection:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
```
nit: Definitions for both idle and idleConnection are same..

nithyatsu commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+The solution would update the specification for the gateway resource in the bicep file to include a timeoutPolicy object. This object would contain the following properties:
+  **response:** The amount of time the gateway should wait for a response from the application before timing out.
+  **idle:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
+  **idleConnection:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
```
idle and idleConnection seem to have same meaning, should we clarify it better when we write the user docs?

superbeeny replied:
Will update

smatochkin commented:
Any thoughts on the future maintenance of the spec with the possible gateway implementation changes? The suggested contract change is tied to Contour specific implementation of the gateway. If there is a plan to switch from Contour specific API to more generic GatewayAPI, then the suggested contact will be hard to migrate.

brooke-hamilton reviewed (commented):
🚀 This is good - agree with checking on the K8s gateway api, as well.

superbeeny commented:
Adding a new comment for timeouts comparison with the gateway api specification. [Route Timeouts](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteTimeouts) have 2 optional fields:  

`request` defined as 

Request specifies the maximum duration for a gateway to respond to an HTTP request. If the gateway has not been able to respond before this deadline is met, the gateway MUST return a timeout error.

and `backendRequest` defined as

BackendRequest specifies a timeout for an individual request from the gateway to a backend. This covers the time from when the request first starts being sent from the gateway to when the full response has been received from the backend.

Is the consensus to update to match the gateway specification?

superbeeny commented:
Can I get a review please @kachawla , @zachcasper @sylvainsf

willdavsmith reviewed (dismissed):
LGTM, great doc!

superbeeny commented:
@willtsai is anything else needed here to approve and merge?

kachawla commented on resources/2025-01-gateway-timeouts/radius-timeout-arch.png:
In this scenario deployment to Kubernetes is handled by Application RP instead of DE.

kachawla commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,350 @@
+
+These proposed changes should not be breaking changes and should not require any changes to existing applications.
+
+#### Disadvantages (of each option considered)
```
Any downsides of this approach long term?

superbeeny replied:
Added notes

kachawla commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,350 @@
+}
+
+@doc("Gateway route timeout policy")
+model GatewayRouteTimeoutPolicy {
+  @doc("The response timeout in duration for the route. Defaults to 15 seconds.")
+  response?: string;
+  
+  @doc("The backend timeout in duration for the route. Cannot be more than the request timeout")
+  backendrequest?: string;
+
+  
+}
```
This doesn't match the schema above in the user experience section, that has `idle` and `idleConnection` instead of `backendrequest`.. is one of the final version? Would also helpful to know why.

superbeeny replied:
request/backendrequest is the final version as it maps directly to the gateway api spec

kachawla commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,350 @@
+-->
+
+The solution would update the specification for the gateway resource in the bicep file to include a timeoutPolicy object. This object would contain the following properties:  
+  **request:** The amount of time the gateway should wait for a response from the application before timing out.  
+  **backendrequest:** The amount of time the gateway should wait for a response from the application before timing out when the connection is idle.
```
Couple of questions -

* What is the difference between the two?
* The property names are different in the subsequent sections of this doc.

superbeeny replied:
request is the overall timeout and backendrequest only applies to downstream connections

kachawla commented:
Thanks for the updates @superbeeny. Did you also get a chance to look into https://github.com/radius-project/design-notes/pull/78#discussion_r1941767274?

kachawla commented:
> Request specifies the maximum duration for a gateway to respond to an HTTP request. If the gateway has not been able to respond before this deadline is met, the gateway MUST return a timeout err

Just saw your comment above on this https://github.com/radius-project/design-notes/pull/78#issuecomment-2637857808. It makes sense to me, thanks for looking into this. Some of the places in the doc need to be updated to reflect this.

superbeeny commented:
Left another comment on #78

kachawla commented:
> Left another comment on #78

Thanks @superbeeny! Let me know when the PR is ready for re-review. This is close to approval :)

superbeeny commented:
@kachawla I think this is ready now, let's have a run through it if not

kachawla commented:
> @kachawla I think this is ready now, let's have a run through it if not

@superbeeny, doesn't look like your updates are reflected in the PR. Could you please look into the comments above from 2 weeks ago?

superbeeny commented:
> > @kachawla I think this is ready now, let's have a run through it if not
> 
> @superbeeny, doesn't look like your updates are reflected in the PR. Could you please look into the comments above from 2 weeks ago?

addressed comments and ready for review

kachawla commented on resources/2025-01-gateway-timeouts.md:
```diff
@@ -0,0 +1,352 @@
+-->
+The goal of this feature is to allow the user to configure a timeout on the gateway for an application within the applications bicep file.
+
+### Non goals
+Other forms of routing configuration are not in scope for this feature. This includes but is not limited to:
+* Configuring timeouts for other types of resources
+* Configuring timeouts for other types of routing
+* Configuring timeouts for other types of applications
+* Configuring timeouts for other types of gateways
+
+<!--
+Describe non-goals to identify something that we won’t be focusing on 
+immediately. We won’t be expending any effort on these matters. If there
+will be follow-ups after this work, list them here. If there are things
+we plan to do in the future, but are out of scope of this design, list
+them here. Provide a brief explanation on why this is a non-goal.
+-->
+None
```
```suggestion
### Non goals
Other forms of routing configuration are not in scope for this feature. This includes but is not limited to:
* Configuring timeouts for other types of resources
* Configuring timeouts for other types of routing
* Configuring timeouts for other types of applications
* Configuring timeouts for other types of gateways
```
